text/0000-injection-hook-normalization.md

```diff
@@ -0,0 +1,269 @@
+[post-creation and pre-destruction][1] hooks for setting up the instance and
+tearing it down after all injections have been assigned.
+
+[1]: https://docs.spring.io/spring/docs/current/spring-framework-reference/core.html#beans-postconstruct-and-predestroy-annotations
+
+This RFC proposes that we normalize the assignment of the owner, and access to
+injected properties, to follow the same conventions as other popular DI
+frameworks. Injections will be assigned _after_ objects have been constructed,
+and a `AFTER_INJECTION` symbol hook will be triggered after that. This will be
+_approximately_ the same flow as `init` is currently for native classes, but is
+a new hook to help with teaching purposes, since `init` has historically been
+taught as the same as `constructor` in native classes (even though it is not).
+We will also trigger a `BEFORE_DESTRUCTION` symbol hook when the container is
+cleaning up objects. Finally, we'll create decorators that users can use to
+decorate functions to run during these lifecycle hooks, so users have a friendly
+and flexible way to define their class's lifecycles.
+
+Doing this will ensure that users can rely on a consistent API for all DI
+related classes, and that the lifecycle of the DI system will be teachable and
+consisent. It also unblocks the usage of plain native classes which do not
+extend a base class in the future. Finally , it will allow us to continue to
+build on the DI system, and
+
+## Detailed design
+
+There are three parts to this design:
+
+### Symbols and Decorators
+
+Two symbols will be added to Ember:
+
+- `AFTER_INJECTION`
+- `BEFORE_DESTRUCTION`
+
+These symbols will be publicly accessible, and will be the keys for the methods
+that should be called on objects at the respective points in their lifecycles.
+We will also add two decorators:
+
+- `@afterInjection`
+- `@beforeDestruction`
+
+These decorators will be usable on methods, and will cause the methods to run
+after injection and before destruction. Multiple methods can be decorated, and
+will all run in the order that decorators are evaluated (from top to bottom in
+the stage 1 spec).
+
+### Glimmer Components
+
+We'll update the `GlimmerComponent` class to _not_ receive any constructor
+parameters. Instead, both `owner` and `args` will be assigned _after_ the object
+has been fully constructed. We will also add the `AFTER_INJECTION` hook, which
+will trigger immediately after `owner` and `args` have been assigned. The
+`willDestroy` hook will also be renamed to `BEFORE_DESTRUCTION`. No default
+implementation will be provided on the base class.
+
+### `EmberObject` based classes (Route/Controller/Service/Helper)
+
+In the case of `EmberObject` classes, `AFTER_INJECTION` will trigger _after_
+`init`. Since the `owner` and arguments are assigned before `init`, it won't be
+necessary to add any assignment code for these.
+
+Importantly, _only_ classes that are constructed by the container will have
```

Comment:
what is `.create` is manually called but [`ownerInjection`](https://api.emberjs.com/ember/3.8/classes/ApplicationInstance/methods/ownerInjection?anchor=ownerInjection)  is provided?

Reply:
If someone manually creates a class like this, then it is not being created by the container, so it will not trigger the conatiner lifecycle hooks